plugins/woocommerce/src/Internal/ProductQueryFilters/FilterClausesGenerator.php

```diff
@@ -254,6 +112,96 @@ private function price_filter_clauses( $args, $wp_query ) {
+	 *
+	 *     @type array {$taxonomy: Attribute taxonomy name} {
+	 *         @type string[] $terms      Chosen terms' slug.
+	 *         @type string   $query_type Query type. Accepts 'and' or 'or'.
+	 *     }
+	 * }
+	 * @return array
+	 */
+	public function add_attribute_clauses( $args, $chosen_attributes ) {
+		if ( empty( $chosen_attributes ) ) {
+			return $args;
+		}
+
+		global $wpdb;
+
+		// The extra derived table ("SELECT product_or_parent_id FROM") is needed for performance
+		// (causes the filtering subquery to be executed only once).
+		$clause_root = " {$wpdb->posts}.ID IN ( SELECT product_or_parent_id FROM (";
+		if ( 'yes' === get_option( 'woocommerce_hide_out_of_stock_items' ) ) {
+			$in_stock_clause = ' AND in_stock = 1';
+		} else {
+			$in_stock_clause = '';
+		}
+
+		$attribute_ids_for_and_filtering = array();
+
+		foreach ( $chosen_attributes as $taxonomy => $data ) {
+			$all_terms                  = get_terms( $taxonomy, array( 'hide_empty' => false ) );
+			$term_ids_by_slug           = wp_list_pluck( $all_terms, 'term_id', 'slug' );
+			$term_ids_to_filter_by      = array_values( array_intersect_key( $term_ids_by_slug, array_flip( $data['terms'] ) ) );
+			$term_ids_to_filter_by      = array_map( 'absint', $term_ids_to_filter_by );
+			$term_ids_to_filter_by_list = '(' . join( ',', $term_ids_to_filter_by ) . ')';
+			$is_and_query               = 'and' === strtolower( $data['query_type'] );
+
+			$count = count( $term_ids_to_filter_by );
+
+			if ( 0 !== $count ) {
+				if ( $is_and_query && $count > 1 ) {
+					$attribute_ids_for_and_filtering = array_merge( $attribute_ids_for_and_filtering, $term_ids_to_filter_by );
+				} else {
+					$clauses[] = "
+							{$clause_root}
+							SELECT product_or_parent_id
+							FROM {$this->get_lookup_table_name()} lt
+							WHERE term_id in {$term_ids_to_filter_by_list}
+							{$in_stock_clause}
+						)";
+				}
+			}
+		}
+
+		if ( ! empty( $attribute_ids_for_and_filtering ) ) {
+			$count                      = count( $attribute_ids_for_and_filtering );
+			$term_ids_to_filter_by_list = '(' . join( ',', $attribute_ids_for_and_filtering ) . ')';
+			$clauses[]                  = "
+				{$clause_root}
+				SELECT product_or_parent_id
+				FROM {$this->get_lookup_table_name()} lt
+				WHERE is_variation_attribute=0
+				{$in_stock_clause}
+				AND term_id in {$term_ids_to_filter_by_list}
+				GROUP BY product_id
+				HAVING COUNT(product_id)={$count}
+				UNION
+				SELECT product_or_parent_id
+				FROM {$this->get_lookup_table_name()} lt
+				WHERE is_variation_attribute=1
+				{$in_stock_clause}
+				AND term_id in {$term_ids_to_filter_by_list}
+				GROUP BY product_or_parent_id
+				HAVING COUNT(product_or_parent_id)>={$count}
+			)";
+		}
+
+		if ( ! empty( $clauses ) ) {
+			// "temp" is needed because the extra derived tables require an alias.
+			$args['where'] .= ' AND (' . join( ' temp ) AND ', $clauses ) . ' temp ))';
+		} elseif ( ! empty( $chosen_attributes ) ) {
```

Comment:
Working on the unit test help me finding [this bug](https://github.com/woocommerce/woocommerce/blob/f2b77b1ada50126a8509704331b8e2eebea03e74/plugins/woocommerce/src/Blocks/QueryFilters.php#L478), phew 😅